### PR TITLE
Cleanup unused guest accounts older than 90 days

### PR DESCRIPTION
**Pull request by bion**
### Description

When the server receives a request from an account with no session, it automatically creates a guest account and associates it with the session returned in the response. This result unused accounts getting created and sitting in the database indefinitely.

This change introduces a new recurring job that deletes accounts and their associated applicant record when they (all of):

- are older than 90 days
- have not authenticated with an identity provider
- have no applications

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >

### Issue(s) this completes

Fixes https://github.com/civiform/civiform/issues/5542, https://github.com/civiform/civiform/issues/4719

### CODE REVIEW

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]
Deleting these unused guest accounts is great. Since there is no actual data I don't see any retention issues from my end. If it's possible other municipalities have different requirements in keeping even empty records we may need to make this something that can be disabled for each deployment.

File: server/app/repository/AccountRepository.java [hunk 1]
```diff
@@ -301,4 +302,33 @@ public ImmutableSet<Account> getGlobalAdmins() {
...
+            + " days' "
+            + "), "
+            + "applicants_deleted AS ("
+            + "  DELETE FROM applicants"
+            + "  WHERE applicants.id IN (SELECT applicant_id FROM unused_accounts) "
+            + ") "
+            + "DELETE FROM accounts "
+            + "WHERE accounts.id IN (SELECT account_id FROM unused_accounts);";
+
+    return database.sqlUpdate(sql).execute();
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/repository/AccountRepository.java:332]: It's all done in a single statement, which creates an implicit transaction in postgres: https://www.postgresql.org/docs/current/tutorial-transactions.html#:~:text=PostgreSQL%20actually%20treats%20every%20SQL,sometimes%20called%20a%20transaction%20block.

File: server/app/repository/AccountRepository.java [hunk 2]
```diff
@@ -301,4 +302,33 @@ public ImmutableSet<Account> getGlobalAdmins() {
...
+            + "  AND applicants.when_created < CURRENT_DATE - INTERVAL '"
+            + maxAgeInDays
+            + " days' "
+            + "), "
+            + "applicants_deleted AS ("
+            + "  DELETE FROM applicants"
+            + "  WHERE applicants.id IN (SELECT applicant_id FROM unused_accounts) "
+            + ") "
+            + "DELETE FROM accounts "
+            + "WHERE accounts.id IN (SELECT account_id FROM unused_accounts);";
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/repository/AccountRepository.java:330]: My intention here is to do both deletes in a single statement, having `applicants_deleted` is just an artifact of trying to do that. How would using a temp table work?

**Comment by bion** [comment]
@gwendolyngoetz 

> If it's possible other municipalities have different requirements in keeping even empty records we may need to make this something that can be disabled for each deployment.

Sent an email asking about this.

File: server/app/repository/AccountRepository.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/repository/AccountRepository.java:330]: I think what was throwing me off was difference between how older version of Microsoft Sql Server CTEs worked. I see [Postgres docs](https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-MODIFYING) having a similar example to what you have so not an issue anymore.

-----
Providing the rest to answer the question of how a temp table would work

```sql
SELECT applicants.account_id AS account_id, applicants.id AS applicant_id 
INTO TEMP unused_accounts
FROM applicants
LEFT JOIN applications ON applicants.id = applications.applicant_id 
LEFT JOIN accounts ON accounts.id = applicants.account_id 
WHERE applications.applicant_id IS NULL 
AND accounts.authority_id IS NULL 
AND applicants.when_created < CURRENT_DATE - INTERVAL '90 days';

DELETE FROM applicants
WHERE applicants.id IN (SELECT applicant_id FROM unused_accounts);
	
DELETE FROM accounts 
WHERE accounts.id IN (SELECT account_id FROM unused_accounts);
```

**Comment by bion** [comment]
Have gotten written approval from civic entities

**Review by gwendolyngoetz — APPROVED** [review]

File: server/app/durablejobs/jobs/UnusedAccountCleanupJob.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import java.time.temporal.ChronoUnit;
+import models.PersistedDurableJob;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import repository.AccountRepository;
+
+/** Destroys all guest accounts older than a set age that have not started any applications. */
+public final class UnusedAccountCleanupJob extends DurableJob {
+  private static final Logger LOGGER = LoggerFactory.getLogger(UnusedAccountCleanupJob.class);
+  private static final int UNUSED_ACCOUNT_MAX_AGE_IN_DAYS = 90;
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/durablejobs/jobs/UnusedAccountCleanupJob.java:16]: great point 😂

**Pull request merged**